### PR TITLE
bus-mapping: fix return_data

### PR DESCRIPTION
**Pull request by pinkiebell**
Return-data wasn't stored correctly or at all on return.
Now we always set `return_data` in the common `handle_return` method.

### CODE REVIEW

File: bus-mapping/src/circuit_input_builder/input_state_ref.rs [hunk 1]
```diff
@@ -796,6 +796,27 @@ impl<'a> CircuitInputStateRef<'a> {
...
+                    // At the moment it conflicts with `call_ctx` and `caller_ctx`.
+                    let callee_memory = self.call_ctx()?.memory.clone();
+                    let caller_ctx = self.caller_ctx_mut()?;
+                    caller_ctx.return_data.resize(length, 0);
+                    caller_ctx.return_data[0..length]
+                        .copy_from_slice(&callee_memory.0[offset..offset + length]);
+                }
+                _ => {
+                    let caller_ctx = self.caller_ctx_mut()?;
+                    caller_ctx.return_data.truncate(0);
```
[the remark this answers is not in the document]
Comment (han0110) [reply bus-mapping/src/circuit_input_builder/input_state_ref.rs:815]: Revisited instructions again, I think so.

**Review by han0110 — APPROVED** [review]
LGTM! So previously the issue is the return data of `REVERT` is not stored, so in circuit input builder when `RETURNDATACOPY` we hit a copy out of bound error, right?

**Comment by pinkiebell** [comment]
> Hey @pinkiebell, could you provide a description of the problem and the fix?
> Just even if is a oneliner, will help. Thanks!

Done. Sorry for silence 😅

**Comment by pinkiebell** [comment]
> LGTM! So previously the issue is the return data of `REVERT` is not stored, so in circuit input builder when `RETURNDATACOPY` we hit a copy out of bound error, right?

Yes, that was the root issue. And it was also the case for `STOP` and friends had stall data from the previous call.

File: bus-mapping/src/circuit_input_builder/input_state_ref.rs [hunk 2]
```diff
@@ -796,6 +796,27 @@ impl<'a> CircuitInputStateRef<'a> {
...
+                OpcodeId::RETURN | OpcodeId::REVERT => {
+                    let offset = step.stack.nth_last(0)?.as_usize();
+                    let length = step.stack.nth_last(1)?.as_usize();
+                    // TODO: Try to get rid of clone.
+                    // At the moment it conflicts with `call_ctx` and `caller_ctx`.
+                    let callee_memory = self.call_ctx()?.memory.clone();
+                    let caller_ctx = self.caller_ctx_mut()?;
+                    caller_ctx.return_data.resize(length, 0);
+                    caller_ctx.return_data[0..length]
+                        .copy_from_slice(&callee_memory.0[offset..offset + length]);
```
[the remark this answers is not in the document]
Comment (pinkiebell) [reply bus-mapping/src/circuit_input_builder/input_state_ref.rs:811]: memory is expanded [here](https://github.com/pinkiebell/zkevm-circuits/blob/f1b39d3ab0d038db319b9711a497e7b68539477e/bus-mapping/src/evm/opcodes/return.rs#L30)

And any exceptions raised in that step should not reach `handle_return` as I know of.

**Review by adria0 — APPROVED** [review]

**Comment by pinkiebell** [comment]
I had to add a zero length check 🙂

**Pull request merged**